### PR TITLE
mon, mgr: move "osd pool stats" command to mgr and mgr python module

### PR DESCRIPTION
**Pull request by Liuchang0812**
I'm going to show those information in ceph-mgr's dashboard, likes recovery_rate / client_io_rate/ degraded objects .. 

@jcsp @liewegas @tchaikov would you mind taking a look?

### CODE REVIEW

File: src/mgr/ActivePyModules.cc [hunk 1]
```diff
@@ -297,6 +297,48 @@ PyObject *ActivePyModules::get_python(const std::string &what)
...
+    string pool_name;
+    PyFormatter f;
+    cluster_state.with_pgmap([&](const PGMap& pg_map) {
+      return cluster_state.with_osdmap([&](const OSDMap& osdmap) {
+	f.open_array_section("pool_stats");
+        for (auto &p : osdmap.get_pools()) {
+	  poolid = p.first;
+	  pool_name = osdmap.get_pool_name(poolid);
+
+	  f.open_object_section("pool");
```
Comment (jcsp) [inline src/mgr/ActivePyModules.cc]: Can you refactor this to share the code between this path and the "osd pool stats" command handling?  This is pretty much the same as all the `if(f)` paths in the other place.

I'm imagining something like a `dump_pool_stats` function on PGMap that takes the poolid and pool_name as an argument.

**Comment by Liuchang0812** [comment]
I found a [issue](http://tracker.ceph.com/issues/22727), it seems that we should not use [delta_sum as pool_stat_t here](https://github.com/ceph/ceph/blob/master/src/mon/PGMap.cc#L394)

```
pool fs_data id 16
  -71/12 objects degraded (-591.667%)
  -874/12 objects misplaced (-7283.333%)
  recovery io 28040 kB/s, 75 objects/s
  client io 85 B/s wr, 0 op/s rd, 0 op/s wr
```

**Comment by jcsp** [comment]
@Liuchang0812 by the way, if you're working on the dashboard piece of this then you might want to rebase since https://github.com/ceph/ceph/pull/19981 merged.

For all these extra details we might want a separate pools page too, rather than trying to squeeze it all on the front page.

**Comment by rjfd** [comment]
@Liuchang0812 now that dashboard_v2 is already in the source tree, could you move you're dashboard code into dashboard_v2?

For the code you have, it looks like you can move your code into a new controller, maybe `controllers/pool.py`?.  Check https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard_v2/HACKING.rst#how-to-add-a-new-controller

You may also ask @tspmelo or @ricardoasmarques to help you in writing the angular2 based frontend code.

**Comment by Liuchang0812** [comment]
@Liuchang0812 now that dashboard_v2 is already in the source tree, could you move you're dashboard code into dashboard_v2?

yeah, of course. I'm going to move it into dashboard_v2. 

thank you all very much.

**Comment by rjfd** [comment]
@Liuchang0812 FYI: I opened a PR that creates the `PoolController` PR #20823

**Comment by LenzGr** [comment]
Any news on that one? It looks as if the dashboard part needs to be rebased and adapted to dashboard v2. Maybe this could be done in a separate PR?

**Comment by Liuchang0812** [comment]
Sorry, @LenzGr. I am too busy recently. There's no any update on this one. Maybe I should push only the first commit in this PR. so that other people could working on this.

**Comment by Liuchang0812** [comment]
@LenzGr done. I'm sorry for the delay on pool page

**Comment by LenzGr** [comment]
@jcsp ping - have your concerns been addressed? As we're unlikely to add support for this in the dashboard for the initial Mimic release, I don't think we need to rush this one in, but your general take on this would be helpful. Thanks!

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-04-14_03:18:44-rados-wip-kefu-testing-2018-04-14-0918-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-04-15_01:40:50-rados-wip-kefu-testing-2018-04-14-0918-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2018-04-15_13:37:16-rados-wip-kefu-testing-2018-04-14-2138-distro-basic-mira/

the failure is addressed by #21433

File: src/mgr/DaemonServer.cc [hunk 2]
```diff
@@ -1050,6 +1050,69 @@ bool DaemonServer::handle_command(MCommand *m)
...
+	    }
+	    one_pool = true;
+	  }
+	  
+	  stringstream rs;
+
+	  if (f)
+	    f->open_array_section("pool_stats");
+	  else
+	  {
```
Comment (tchaikov) [inline src/mgr/DaemonServer.cc]: move the `{` up to the end of previous line.

File: src/mgr/DaemonServer.cc [hunk 3]
```diff
@@ -1050,6 +1050,69 @@ bool DaemonServer::handle_command(MCommand *m)
...
+	    if (osdmap.get_pools().empty())
+	    {
```
Comment (tchaikov) [inline src/mgr/DaemonServer.cc]: ditto.

File: src/mgr/DaemonServer.cc (hunk 2 above)
Comment (Liuchang0812) [reply src/mgr/DaemonServer.cc]: done, with following patch:

```diff
diff --git a/src/mgr/DaemonServer.cc b/src/mgr/DaemonServer.cc
index d6a201bfff..56dedb8c99 100644
--- a/src/mgr/DaemonServer.cc
+++ b/src/mgr/DaemonServer.cc
@@ -1073,17 +1073,14 @@ bool DaemonServer::handle_command(MCommand *m)
 
 	  if (f)
 	    f->open_array_section("pool_stats");
-	  else
-	  {
-	    if (osdmap.get_pools().empty())
-	    {
+	  else {
+	    if (osdmap.get_pools().empty()) {
 	      ss << "there are no pools!";
 	      goto stats_out;
 	    }
 	  }
 
-	  for (auto &p : osdmap.get_pools())
-	  {
+	  for (auto &p : osdmap.get_pools()) {
 	    if (!one_pool) {
 	      poolid = p.first;
             }
@@ -1096,13 +1093,11 @@ bool DaemonServer::handle_command(MCommand *m)
 	  }
 
 	stats_out:
-	  if (f)
-	  {
+	  if (f) {
 	    f->close_section();
 	    f->flush(cmdctx->odata);
 	  }
-	  else
-	  {
+	  else {
 	    cmdctx->odata.append(rs.str());
 	  }
 	  return 0;
diff --git a/src/mon/PGMap.cc b/src/mon/PGMap.cc
index 24816094d8..8501fa3ce6 100644
--- a/src/mon/PGMap.cc
+++ b/src/mon/PGMap.cc
@@ -2123,8 +2123,7 @@ void PGMap::dump_filtered_pg_stats(ostream& ss, set<pg_t>& pgs) const
 void PGMap::dump_pool_stats_and_io_rate(int64_t poolid, const OSDMap &osd_map, Formatter *f, stringstream *rs) const {
   string pool_name = osd_map.get_pool_name(poolid);
 
-  if (f)
-  {
+  if (f) {
     f->open_object_section("pool");
     f->dump_string("pool_name", pool_name.c_str());
     f->dump_int("pool_id", poolid);
@@ -2134,14 +2133,12 @@ void PGMap::dump_pool_stats_and_io_rate(int64_t poolid, const OSDMap &osd_map, F
   list<string> sl;
   stringstream tss;
   pool_recovery_summary(f, &sl, poolid);
-  if (!f && !sl.empty())
-  {
+  if (!f && !sl.empty()) {
     for (auto &p : sl)
       tss << "  " << p << "\n";
   }
 
-  if (f)
-  {
+  if (f) {
     f->close_section(); // object section recovery
     f->open_object_section("recovery_rate");
   }
@@ -2151,8 +2148,7 @@ void PGMap::dump_pool_stats_and_io_rate(int64_t poolid, const OSDMap &osd_map, F
   if (!f && !rss.str().empty())
     tss << "  recovery io " << rss.str() << "\n";
 
-  if (f)
-  {
+  if (f) {
     f->close_section(); // object section recovery_rate
     f->open_object_section("client_io_rate");
   }
@@ -2165,10 +2161,8 @@ void PGMap::dump_pool_stats_and_io_rate(int64_t poolid, const OSDMap &osd_map, F
 
   // dump cache tier IO rate for cache pool
   const pg_pool_t *pool = osd_map.get_pg_pool(poolid);
-  if (pool->is_tier())
-  {
-    if (f)
-    {
+  if (pool->is_tier()) {
+    if (f) {
       f->close_section(); // object section client_io_rate
       f->open_object_section("cache_io_rate");
     }
@@ -2179,8 +2173,7 @@ void PGMap::dump_pool_stats_and_io_rate(int64_t poolid, const OSDMap &osd_map, F
     if (!f && !rss.str().empty())
       tss << "  cache tier io " << rss.str() << "\n";
   }
-  if (f) 
-  {
+  if (f) {
     f->close_section(); // object section cache_io_rate
     f->close_section(); // object section pool
   } else {
```

File: src/mgr/DaemonServer.cc [hunk 4]
```diff
@@ -1089,6 +1089,64 @@ bool DaemonServer::handle_command(MCommand *m)
...
+	      break;
+	    }
+	  }
+
+	stats_out:
+	  if (f) {
+	    f->close_section();
+	    f->flush(cmdctx->odata);
+	  }
+	  else {
```
Comment (tchaikov) [inline src/mgr/DaemonServer.cc]: please move `else {` to the previous line.

File: src/mgr/DaemonServer.cc [hunk 5]
```diff
@@ -1089,6 +1089,64 @@ bool DaemonServer::handle_command(MCommand *m)
...
+	    cmdctx->odata.append(rs.str());
+	  }
+	  return 0;
+      });
+    });
+    if (r != -EOPNOTSUPP) {
+      cmdctx->reply(r, ss);
+      return true;
+    }
+
```
Comment (tchaikov) [inline src/mgr/DaemonServer.cc]: please remove this empty line.

File: src/mon/PGMap.cc [hunk 6]
```diff
@@ -2120,6 +2120,71 @@ void PGMap::dump_filtered_pg_stats(ostream& ss, set<pg_t>& pgs) const
   ss << tab;
 }
 
+void PGMap::dump_pool_stats_and_io_rate(int64_t poolid, const OSDMap &osd_map, Formatter *f, stringstream *rs) const {
```
Comment (tchaikov) [inline src/mon/PGMap.cc]: please wrap this line at 79 chars.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-04-20_05:42:59-rados-wip-kefu-testing-2018-04-20-1141-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2018-04-20_14:33:42-rados-wip-kefu-testing-2018-04-20-1141-distro-basic-mira/

the failures are addressed by 

- #21549
- #21579

**Comment by tchaikov** [comment]
retest this please.

**Pull request merged**